src/CsharpClient/QuixStreams.Kafka.Transport/SerDes/KafkaMessageMerger.cs

```diff
@@ -0,0 +1,212 @@
+﻿using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Extensions.Logging;
+using QuixStreams.Kafka.Transport.SerDes.Legacy;
+
+namespace QuixStreams.Kafka.Transport.SerDes
+{
+    /// <summary>
+    /// 
+    /// </summary>
+    public class KafkaMessageMerger
+    {
+        private readonly ILogger logger = QuixStreams.Logging.CreateLogger<KafkaMessageMerger>();
+
+        private long bufferCounter = 0;
+        private readonly ConcurrentDictionary<MergerBufferId, PendingKafkaMessage> pendingMessages = new ConcurrentDictionary<MergerBufferId, PendingKafkaMessage>(); // Messages that are queued up
+        private readonly SortedDictionary<long, MergerBufferId> packageOrder = new SortedDictionary<long, MergerBufferId>(); // the order the packages should be raised
+
+        private readonly SemaphoreSlim semaphoreSlim = new SemaphoreSlim(1,1); // see https://blog.cdemi.io/async-waiting-inside-c-sharp-locks/
+        private readonly KafkaMessageMergerHelper helper;
+        /// <summary>
+        /// Initializes a new instance of <see cref="KafkaMessageMerger"/>
+        /// </summary>
+        public KafkaMessageMerger(KafkaMessageBuffer messageBuffer)
+        {
+            this.helper = new KafkaMessageMergerHelper(messageBuffer, this.logger);
+            this.helper.OnMessageSegmentsPurged += (bufferId) =>
+            {
+                if (this.RemoveFromBuffer(bufferId))
+                {
+                    RaiseNextPackageIfReady().GetAwaiter().GetResult();
+                }
+            };
```

Comment:
I think this should also be possible
```suggestion
            this.helper.OnMessageSegmentsPurged += async (bufferId) =>
            {
                if (this.RemoveFromBuffer(bufferId))
                {
                    await RaiseNextPackageIfReady();
                }
            };
```

Reply:
On first glance it looks like it could be done, but it has quite the cascading effect up the chain, making a lot of the code async or do the same thing I did here. After some investigation I decided to not go for this for now.